### PR TITLE
SQL-195 Async Reaction Processing

### PR DESCRIPTION
**Pull request by milt**
[SQL-195] Asynchronously react to statement submission.

[SQL-195]: https://yet.atlassian.net/browse/SQL-195?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: doc/env_vars.md [hunk 1]
```diff
@@ -104,6 +104,8 @@ The following options are used for advanced database performance tuning and may
...
 | `LRSQL_STMT_RETRY_BUDGET`       | `stmtRetryBudget`       | The max amount of time allowed for statement POST transaction retries before failing (ms).                                                                                                                        | `1000`                                                                                |
+| `LRSQL_ENABLE_REACTIONS`        | `enableReactions`       | Whether or not to enable statement reactions.                                                                                                                                                                     | `false`                                                                               |
+| `LRSQL_REACTION_BUFFER_SIZE`    | `reactionBufferSize`    | Number of pending reactions to allow.                                                                                                                                                                             | `10000`                                                                               |
```
Comment (cliffcaseyyet) [inline doc/env_vars.md:108]: Add some comment here about what happens to overflow.

File: src/main/lrsql/init/reaction.clj [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+(defn shutdown-reactions!
+  "Given a (possibly nil) reaction channel and reaction executor channel,
+  gracefully shut them down."
+  [?reaction-channel
+   ?reaction-executor]
+  (when (and ?reaction-channel ?reaction-executor)
+    (log/info "Stopping reaction processor...")
+    (log/debug "Closing reaction channel...")
+    (a/close! ?reaction-channel)
+    (log/debug "Draining reaction buffer...")
```
Comment (cliffcaseyyet) [inline src/main/lrsql/init/reaction.clj:86]: dope

File: src/main/lrsql/system/lrs.clj [hunk 3]
```diff
@@ -136,8 +142,11 @@
                     stmt-result)
                 ;; Non-error result - continue
                 (if-some [stmt-id (:statement-id stmt-result)]
-                  (recur (rest stmt-ins)
-                         (update stmt-res :statement-ids conj stmt-id))
+                  (do
+                    ;; Submit statement for reaction if enabled
+                    (react-init/offer-trigger! reaction-channel stmt-id)
+                    (recur (rest stmt-ins)
```
Comment (cliffcaseyyet) [inline src/main/lrsql/system/lrs.clj:148]: what is `rest` doing in this case? Ensuring that insert has completed? Is it lazy?

**Review by cliffcaseyyet — COMMENTED** [review]
Incredibly straightforward to read. Looks really good. Usually your async stuff throws me but I guess this is a simple enough processing loop.

File: src/main/lrsql/init/reaction.clj [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+               :trigger-id ::xs/uuid)
+  :ret nil?)
+
+(defn offer-trigger!
+  "Given a (possibly nil) reaction channel and a string statement ID, submit the
+  ID to the channel as a UUID if it exists, or do nothing if it is nil.
+  Log if the channel exists but the ID cannot be submitted."
+  [?reaction-channel trigger-id]
+  (when ?reaction-channel
+    (when-not (a/offer! ?reaction-channel (u/str->uuid trigger-id))
```
[the remark this answers is not in the document]
Comment (milt) [reply src/main/lrsql/init/reaction.clj:37]: When reactions are off there is no reaction channel and it doesn't do anything. I chose this structure to keep conditionals out of the component ns to keep it simple.

File: src/main/lrsql/init/reaction.clj (hunk 2 above)
Comment (milt) [reply src/main/lrsql/init/reaction.clj:86]: Hopefully noisy enough but we'll see in debugging

File: src/main/lrsql/system/lrs.clj (hunk 3 above)
Comment (milt) [reply src/main/lrsql/system/lrs.clj:148]: That's part of the existing code here, it loops through the sequence of statements, storing them and keeping the ids for novel ones

File: doc/env_vars.md (hunk 1 above)
Comment (milt) [reply doc/env_vars.md:108]: Will do. Also gonna make a separate doc page on reactions outside this pr.

Comment (milt) [reply doc/env_vars.md]: https://github.com/yetanalytics/lrsql/pull/316/commits/689f9540c19f4d0ca4f77a51e2baa3cd746a3d19

**Review by cliffcaseyyet — APPROVED** [review]

**Pull request merged**